exporter/cassandra/exporter.go

```diff
@@ -0,0 +1,164 @@
+package cassandra
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/translator/jaeger"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumererror"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger-opentelemetry-collector/internal/metrics"
+	"github.com/jaegertracing/jaeger/pkg/cassandra"
+	casConfig "github.com/jaegertracing/jaeger/pkg/cassandra/config"
+	plugin "github.com/jaegertracing/jaeger/plugin/storage/cassandra"
+	cSpanStore "github.com/jaegertracing/jaeger/plugin/storage/cassandra/spanstore"
+	"github.com/jaegertracing/jaeger/plugin/storage/cassandra/spanstore/dbmodel"
+)
+
+func newTracesExporter(cfg *Config, set component.ExporterCreateSettings) (component.TracesExporter, error) {
+	s := newSender(cfg, set.TelemetrySettings)
+	return newTracesExporterWithSender(cfg, set, s)
+}
+
+func newTracesExporterWithSender(cfg *Config, set component.ExporterCreateSettings, sender *sender) (component.TracesExporter, error) {
+	return exporterhelper.NewTracesExporter(
+		context.TODO(), set, cfg, sender.pushTraces,
+		exporterhelper.WithCapabilities(consumer.Capabilities{MutatesData: false}),
+		exporterhelper.WithStart(sender.start),
+		exporterhelper.WithShutdown(sender.shutdown),
+		exporterhelper.WithTimeout(cfg.TimeoutSettings),
+		exporterhelper.WithRetry(cfg.RetrySettings),
+		exporterhelper.WithQueue(cfg.QueueSettings),
+	)
+}
+
+type sender struct {
+	name           string
+	settings       component.TelemetrySettings
+	cfg            *Config
+	primaryConfig  casConfig.SessionBuilder
+	primarySession cassandra.Session
+	// factory  *plugin.Factory // Cannot use unless some functions or fields are exported
+	writer *cSpanStore.SpanWriter
+}
+
+func newSender(cfg *Config, settings component.TelemetrySettings) *sender {
+	return &sender{
+		name:          cfg.ID().String(),
+		settings:      settings,
+		cfg:           cfg,
+		primaryConfig: cfg.Options.GetPrimary(),
+	}
+}
+
+func (s *sender) pushTraces(ctx context.Context, td ptrace.Traces) error {
+	batches, err := jaeger.ProtoFromTraces(td)
+	if err != nil {
+		return consumererror.NewPermanent(fmt.Errorf("failed to push trace data via cassandra exporter: %w", err))
+	}
+
+	for _, batch := range batches {
+		process := batch.GetProcess()
+		if process == nil {
+			s.settings.Logger.Debug("NO PROCESS")
+		}
+		for _, span := range batch.GetSpans() {
+			span.Process = process
+			err = s.writer.WriteSpan(ctx, span)
```

Comment:
My novice understanding is that that's done for me via the mechanisms available in [exporterhelper](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter/exporterhelper).

The queue actually sets up a set of workers to dequeue, 10 by default. So the function is synchronous here but there's many workers. The incoming `td` may or may not be a batch, it could just be a single span. It depends upon if somewhere earlier in the write path the batch processor is used. But I use batch here as the translator is returning a batch defined in the Jaeger model https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c0666b60f669993fc07369c60d3d7c13143df599/pkg/translator/jaeger/traces_to_jaegerproto.go#L29.

The retry helper takes the full 'batch' of failed spans and attempts to retry.

Ah there's my bug. If the first succeeds and the second doesn't the first would be rewritten again. Which I guess doesn't matter, the data would still be consistent (it would just overwrite) but we waste a write. I'd think you'd have to have a de-batcher earlier in the pipeline to address that issue.